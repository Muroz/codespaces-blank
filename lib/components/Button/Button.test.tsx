import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';

import { Button } from '.';

const buttonContent = 'Action';

describe('Button', () => {
  it('renders correctly with the default variant', () => {
    const { container } = render(<Button>{buttonContent}</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with the outline variant', () => {
    const { container } = render(
      <Button variant="outline">{buttonContent}</Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with the ghost variant', () => {
    const { container } = render(
      <Button variant="ghost">{buttonContent}</Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders without accessibility violations for the default variant', async () => {
    const { container } = render(<Button>{buttonContent}</Button>);
    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('renders without accessibility violations for the outline variant', async () => {
    const { container } = render(
      <Button variant="outline">{buttonContent}</Button>
    );
    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('renders without accessibility violations for the ghost variant', async () => {
    const { container } = render(
      <Button variant="ghost">{buttonContent}</Button>
    );
    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('triggers the expected events on button click', async () => {
    const handleOnClick = vi.fn();
    const handleOnMouseDown = vi.fn();
    const handleOnMouseUp = vi.fn();

    render(
      <Button
        onClick={handleOnClick}
        onMouseDown={handleOnMouseDown}
        onMouseUp={handleOnMouseUp}
      >
        {buttonContent}
      </Button>
    );
    const user = userEvent.setup();

    const button = screen.getByRole('button', { name: buttonContent });

    await user.click(button);

    expect(handleOnClick).toHaveBeenCalledTimes(1);
    expect(handleOnMouseDown).toHaveBeenCalledTimes(1);
    expect(handleOnMouseUp).toHaveBeenCalledTimes(1);
  });
});
