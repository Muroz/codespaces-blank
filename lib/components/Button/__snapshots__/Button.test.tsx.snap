// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > renders correctly with the default variant 1`] = `
.c0 {
  border-color: transparent;
  border-width: 1px;
  margin: 0.5rem;
  font-size: 1.5rem;
  letter-spacing: 1px;
  font-weight: 700;
  line-height: 1.2;
  box-sizing: border-box;
  min-width: 12rem;
  padding: 1rem 2rem;
  background: #0073e6;
  color: #ffffff;
}

.c0:hover {
  background: #0056b8;
}

.c0:focus {
  background: #0073e6;
  outline: 4px solid;
  outline-color: #0073e6;
  outline-offset: 2px;
}

.c0:active {
  background: #003c8a;
}

.c0:disabled,
.c0:hover:disabled,
.c0:hover:focus:disabled {
  background: #a6c3d8;
  color: #466678;
}

<button
  class="c0"
  variant="solid"
>
  Action
</button>
`;

exports[`Button > renders correctly with the ghost variant 1`] = `
.c0 {
  border-color: transparent;
  border-width: 1px;
  margin: 0.5rem;
  font-size: 1.5rem;
  letter-spacing: 1px;
  font-weight: 700;
  line-height: 1.2;
  box-sizing: border-box;
  min-width: 12rem;
  padding: 1rem 2rem;
  background: transparent;
  color: #0073e6;
}

.c0:hover {
  background: #e6f2f9;
}

.c0:focus {
  outline: 4px solid;
  outline-color: #0073e6;
  outline-offset: 2px;
}

.c0:active {
  background: #c8deec;
}

.c0:disabled,
.c0:hover:disabled,
.c0:hover:focus:disabled {
  color: #466678;
}

<button
  class="c0"
  variant="ghost"
>
  Action
</button>
`;

exports[`Button > renders correctly with the outline variant 1`] = `
.c0 {
  border-color: transparent;
  border-width: 1px;
  margin: 0.5rem;
  font-size: 1.5rem;
  letter-spacing: 1px;
  font-weight: 700;
  line-height: 1.2;
  box-sizing: border-box;
  min-width: 12rem;
  padding: 1rem 2rem;
  background: transparent;
  border-color: #0073e6;
  border-style: solid;
  border-width: 1px;
  color: #0073e6;
}

.c0:hover {
  background: #0073e6;
  color: #ffffff;
}

.c0:focus {
  background: #0073e6;
  color: #ffffff;
  outline: 4px solid;
  outline-color: #0073e6;
  outline-offset: 2px;
}

.c0:active {
  background: #003c8a;
  color: #ffffff;
}

.c0:disabled,
.c0:hover:disabled,
.c0:hover:focus:disabled {
  background: transparent;
  color: #466678;
  border-color: #466678;
}

<button
  class="c0"
  variant="outline"
>
  Action
</button>
`;
